src/machine/machine_mimxrt1062_i2c.go

```diff
@@ -0,0 +1,646 @@
+// +build mimxrt1062
+
+package machine
+
+// I2C peripheral abstraction layer for the MIMXRT1062
+
+import (
+	"device/nxp"
+	"errors"
+)
+
+const (
+	TWI_FREQ_BUS     = 24000000        // LPI2C root clock is on 24 MHz OSC
+	TWI_FREQ_100KHZ  = 100000          // StandardMode (100 kHz)
+	TWI_FREQ_400KHZ  = 400000          // FastMode (400 kHz)
+	TWI_FREQ_1MHZ    = 1000000         // FastModePlus (1 MHz)
+	TWI_FREQ_5MHZ    = 5000000         // UltraFastMode (5 MHz)
+	TWI_FREQ_DEFAULT = TWI_FREQ_100KHZ // default to StandardMode (100 kHz)
+)
+
+var (
+	errI2CWriteTimeout       = errors.New("I2C timeout during write")
+	errI2CReadTimeout        = errors.New("I2C timeout during read")
+	errI2CBusReadyTimeout    = errors.New("I2C timeout on bus ready")
+	errI2CSignalStartTimeout = errors.New("I2C timeout on signal start")
+	errI2CSignalReadTimeout  = errors.New("I2C timeout on signal read")
+	errI2CSignalStopTimeout  = errors.New("I2C timeout on signal stop")
+	errI2CAckExpected        = errors.New("I2C error: expected ACK not NACK")
+	errI2CBusError           = errors.New("I2C bus error")
+	errI2CNotConfigured      = errors.New("I2C interface is not yet configured")
+)
+
+// I2CConfig is used to store config info for I2C.
+type I2CConfig struct {
+	Frequency uint32
+	SDA       Pin
+	SCL       Pin
+}
+
+func (c I2CConfig) getPins() (sda, scl Pin) {
+	if 0 == c.SDA && 0 == c.SCL {
+		// default pins if none specified
+		return I2C_SDA_PIN, I2C_SCL_PIN
+	}
+	return c.SDA, c.SCL
+}
```

Comment:
I think Pin needs to be specified according to the value of muxSelect{}.
If I'm right about this, why not add the Pin information to muxSelect{}?

Or, depending on the settings of I2CConfig-SDA, etc., maxSelect{} should be generated automatically.

Reply:
Not sure if I'm understanding this one correctly, but adding pin information to `muxSelect{}` would: 
1. Require another (large) lookup table that would take considerable effort to create, and 
2. The user would still need to define which `muxSelect{}` to associate with their `I2C{}`. 

The intent of this method was to automatically determine the pins in case the user has intentionally left their `I2C{}` _unconfigured_ – to let TinyGo determine the default settings.

There are two separate devices that must be configured to connect a given pin to a given peripheral. For example, digital pin `18` (port `A` pin `PA17`, also known as `AD_B1_01`) can provide I2C SDA functionality to `I2C1` (peripheral `LPI2C1`):
|Alt Fn|GPIO|IOMUXC|
|:-----:|:----:|:-------:|
|I2C SDA|Configure `AD_B1_01` as `LPI2C1 SDA`|Connect `LPI2C1 SDA` to `AD_B1_01`|

The tricky part is while `AD_B1_01` can be configured with different alternate functions (e.g., I2C SDA, SPI MOSI, etc.), **multiple different pins can provide I2C SDA functionality to `LPI2C1`**.

In the code, the GPIO configuration is done in `(Pin).Configure(PinConfig)` via `(Pin).getMuxMode(PinConfig)`, and IOMUXC connection is done in `(*I2C).Configure(I2CConfig)` via this `(I2CConfig).getPins()` method. 

Similar to how `(Pin).getMuxMode(PinConfig)` automatically determines the alternate function mode based on both `Pin` and peripheral (via `PinConfig`), this `(I2CConfig).getPins()` could be expanded to automatically determine the pins based on peripheral alone (we already know the alternate function mode is I2C SDA/SCL). Something _similar_ to this:

```go
func (c I2CConfig) getPins(i2c *I2C) (sda, scl Pin) {
	if 0 == c.SDA && 0 == c.SCL {
		// default pins if none specified
		switch i2c {
		case I2C1:
			return I2C1_SDA_PIN, I2C1_SCL_PIN
		case I2C2:
			return I2C2_SDA_PIN, I2C2_SCL_PIN
		case I2C3:
			return I2C3_SDA_PIN, I2C3_SCL_PIN
		}
		// the given *I2C is unsupported
		return I2C_SDA_PIN, I2C_SCL_PIN
	}
	return c.SDA, c.SCL
}
```

I'll go ahead and update this method if this looks right to you.

Reply:
I've added default SDA, SCL pins to type `I2C`:

```diff
  type I2C struct {
  	Bus *nxp.LPI2C_Type

+  	// these pins are initialized by each global I2C variable declared in the
+  	// board_teensy4x.go file according to the board manufacturer's default pin
+  	// mapping. they can be overridden with the I2CConfig argument given to
+  	// (*I2C).Configure(I2CConfig).
+  	sda, scl Pin
  
  	// these hold the input selector ("daisy chain") values that select which pins
  	// are connected to the LPI2C device, and should be defined where the I2C
  	// instance is declared (e.g., in the board definition). see the godoc
  	// comments on type muxSelect for more details.
  	muxSDA, muxSCL muxSelect
  }
```

And then for each peripheral, added something like:

```diff
  	_I2C1 = I2C{
  		Bus: nxp.LPI2C1,
+  		sda: I2C1_SDA_PIN, // D18 (PA17 [AD_B1_01])
+  		scl: I2C1_SCL_PIN, // D19 (PA16 [AD_B1_00])
-  		muxSDA: muxSelect{ // D18 (PA17 [AD_B1_01])
+  		muxSDA: muxSelect{
  			mux: nxp.IOMUXC_LPI2C1_SDA_SELECT_INPUT_DAISY_GPIO_AD_B1_01_ALT3,
  			sel: &nxp.IOMUXC.LPI2C1_SDA_SELECT_INPUT,
  		},
-  		muxSCL: muxSelect{ // D19 (PA16 [AD_B1_00])
+  		muxSCL: muxSelect{
  			mux: nxp.IOMUXC_LPI2C1_SCL_SELECT_INPUT_DAISY_GPIO_AD_B1_00_ALT3,
  			sel: &nxp.IOMUXC.LPI2C1_SCL_SELECT_INPUT,
  		},
  	}
```

The method `(I2CConfig).getPins() (Pin, Pin)` was replaced with a method to override these default pins when the user calls `(*I2C).Configure(I2CConfig)`:

```diff
+  func (i2c *I2C) setPins(c I2CConfig) (sda, scl Pin) {
+  	// if both given pins are defined, or either receiver pin is undefined.
+  	if 0 != c.SDA && 0 != c.SCL || 0 == i2c.sda || 0 == i2c.scl {
+  		// override the receiver's pins.
+  		i2c.sda, i2c.scl = c.SDA, c.SCL
+  	}
+  	// return the selected pins.
+  	return i2c.sda, i2c.scl
+  }
  
  // Configure is intended to setup an I2C interface for transmit/receive.
  func (i2c *I2C) Configure(config I2CConfig) {
  	// init pins
-  	sda, scl := config.getPins()
+  	sda, scl := i2c.setPins(config)
  
  	// configure the mux and pad control registers
  	sda.Configure(PinConfig{Mode: PinModeI2CSDA})
  	scl.Configure(PinConfig{Mode: PinModeI2CSCL})
  
  	// configure the mux input selector
  	i2c.muxSDA.connect()
  	i2c.muxSCL.connect()
```

Now, whether the user calls `p.Configure(I2CConfig{})` with `p` being `I2C0`, `I2C1`, `I2C2`, `I2C3`, or `I2C4`, it will use the correct pins for the given peripheral by default.